Release 0.9

jamesmontalvo3 opened the pull request:
This pull request bumps meza to version 0.9. The full description of changes and required testing will be completed later.

### Testing

- [ ] `curl -LO https://raw.github.com/enterprisemediawiki/meza/REL0_9/scripts/install.sh`
- [ ] `sudo bash install.sh`
- [ ] `REL0_9` branch
- [ ] Perform tests from [CONTRIBUTING.md](https://github.com/enterprisemediawiki/meza/blob/master/CONTRIBUTING.md)
  1. Minimal requirements
  2. Desired testing
  3. Pre-release testing requirements (minus CentOS 6 requirements, which we do not currently support)
- [ ] Try different settings in `/opt/meza/config/local/LandingPage.php`
- [ ] Try different settings for `$mezaAuthType` in `/opt/meza/config/local/preLocalSettings_allWikis.php` and individual wikis' `preLocalSettings.php` files
- [ ] Try loading other extensions similar to how performed in `/opt/meza/config/template/more-extensions.php`
- [ ] Verify functionality of Elasticsearch API and plugins (See [manual/ElasticsearchPlugins.md](manual/ElasticsearchPlugins.md) for info)
- [ ] Test email functionality (make sure it's enabled in `/opt/meza/config/local/preLocalSettings_allWikis.php`)
- [ ] Unify user tables (run `sudo WIKI=insertanywiki php /opt/meza/scripts/unifyUserTables.php --prime-wiki=yourprimewiki`)
- [ ] Rebuild an Elasticsearch index with `sudo bash elastic-rebuild-index.sh`

### Changes

* Most settings files are now in `/opt/meza/config` and are symlinked from their required locations as required (e.g. `/etc/parsoid/api/localsettings.js` is a symlink to `/opt/meza/config/core/localsettings.js`)
* Add `/opt/meza/config/core/BlenderSettings.php` file with meza requirements. 
  * A template for instance-specific settings is in `/opt/meza/config/template/LandingPage.php` which is copied into `/opt/meza/config/local` during installation
* LocalSettings.php
  * Add "permissions" section to `LocalSettings.php` to handle standard meza permissions. Of course a meza instance can choose to use whatever permissions they want, but this provides some simple defaults.
  * Remove "Homeless Items" from `LocalSettings.php` since our config is getting more stable
  * Made `$mezaCommandLineDebug = true;` by default since we always want debugging on for command line operations to get the most verbose error reporting possible.
  * If the file `/opt/meza/config/local/primewiki` exists, in `LocalSettings.php` read the name of the wiki in that file and make it the prime wiki. All user and interwiki info for all wikis will be stored in that wiki.
  * Don't use memcached for sessions due to conflict with SimpleSamlPhp
  * Remove IMSQuery and SummaryTimeline due to being overly specific to certain installations. They have been move to `/opt/meza/config/template/more-extensions.php` along with the addition of extensions HideSubPage, CrossReference, and TreeAndMenu.
  * Disable PdfHandler extension due to failure on large PDFs
  * Disable the Flow extension due to lack of searchability
  * Enable Extension:CollabsibleVector due to collapsible features being removed from newer versions of the Vector skin.
* Created instance-specific file `/opt/meza/config/local/domain` to hold the domain name or IP address. This should be used everywhere the domain name is required so the domain can easily be edited. At present only `localsettings.js` is using it, I think.
* Apache installed with yum wants the Non-threadsafe version of PHP. This also required changing the version of opcache used in php.ini.
* Create httpd.conf in `/opt/meza/config/core`
  * Add port 9000 entrypoint for Parsoid only (not accessible externally through firewall). This was previously 8142 but SELinux does not want Apache to use that port. We have SELinux in "permissive mode" now anyway, so it doesn't really matter, but the hope is to get back to "enforcing mode" at some point.
  * Add port 8008 as reverse proxy for accessing Elasticsearch API. Initially this was setup so only GET requests could be made (no POST, PUT or DELETE), however this conflicted with many of the ES plugins. For now these are exposed, but we should perhaps consider closing this off later on. That said, the API is not accessible by default and is password protected once configured. See [manual/ElasticsearchPlugins.md](manual/ElasticsearchPlugins.md) for info on setup.
  * Added, but commented out due to issues with Elasticsearch, redirection from http
* `/opt/meza/scripts/commandTimes.js` was created to determine how long each part of the installation took. Additionally, these times will be sent along to a Slack webhook after completion of installation if desired. This is done with new script `/opt/meza/scripts/slack.sh`.
* Many files moved from `/opt/meza/scripts` to `/opt/meza/config`
* Added `/opt/meza/scripts/elastic-rebuild-index.sh` to allow rebuilding Elasticsearch indexes for a given wiki. This script is just a wrapper on `/opt/meza/scripts/elastic-build-index.sh`.
* SMW now checks out a specific commit that is nearly version 2.4 in order to get certain important aspects of 2.4 (Basically rebuildData.php fails on some pages without this. Also better error reporting.). Once 2.4 is released it will be changed to checkout that release, rather than a specific commit. We can't checkout 2.4 because it's not released yet. We'd have to be running on SMW's master branch, which is a moving target (something we've tried to avoid).
* Modified `import-remote-wikis.sh` to be mroe verbose. Use verbose `mysqldump`, and use `rsync` instead of `cp`. Additionally, using `rsync` provides better failure recovery, since it can be run multiple times and pick up where it left off. It also could be used to keep two servers synced.
* Email enabled, requiring the PEAR mail package for ...
* Remove options to pick PHP version and MW API protocol (it's always https now)
* Install more stuff with yum:
  * httpd-devel (apache webserver)
  * mod_ssl (apache module to support https)
  * mod_proxy_html (apache module to support proxies, as required for Elasticsearch plugins)
  * net-tools (provides useful `netstat` command)
  * vim (better text editing)
  * sendmail (for email)
  * sendmail-cf (for email)
  * m4 (for email)
  * expect (used to install PEAR, which requires prompts. Expect can be used to simulate user-entered prompts, and is used during install to make it so we don't have to have the user enter the information)
  * expectk (ditto)
* SAML pre- and post-LocalSettings files updated to disallow account creation and remove login/logout
* SAML setup config file template created at `/opt/meza/config/template/saml-setup-config.sh`. This is to avoid prompts in the SAML setup script.
* Add `/opt/meza/config/template/install.config.example.sh` which can be used to avoid having to enter prompts 
* Better naming of setup files
  * Change `...wikis/$wikiId/config/setup.php` to `preLocalSettings.php`
  * Change `...wikis/$wikiId/config/overrides.php` to `postLocalSettings.php`
  * Add `/opt/meza/config/local/postLocalSettings_allWikis.php`
  * Created standard `/opt/meza/config/template/preLocalSettings_allWikis.php`
  * SSH/networking setup script was just `setup.sh`, which was confusing with all the `setup.php` scripts and `install.sh`. It was renamed `dev-networking.sh` since it's only used in dev VMs.

### Issues and PRs

The following issues and PRs are addressed in this PR into master. See also [all Pull Requests that were merged into REL0_9](https://github.com/enterprisemediawiki/meza/pulls?utf8=✓&q=is%3Apr+base%3AREL0_9+).

* #204: Provide method to declare primary wiki for user table, interwiki, etc
* #205: unifyUserTables.php script to make all wikis use the same user table
* #70: Create rebuild index script
* #347, #356: Fix confusing names like "setup.php"
* #355: make $mezaAuthType do something
* #358: Disable PdfHandler for now, since it breaks wikis with large PDFs
* #363: Rename `/opt/meza/config/meza` to `core`
* #362: Use SMW commit that doesn't fail on rebuildData.php
* #281, #255: Use Elasticsearch management plugins
* #350: Disable memcached for sessions (for now) due to SimpleSamlPhp conflict
* #308: Move all config files to single directory called /config
* #312, #315 Install Apache httpd with Yum
* #313: Bump MW to 1.25.x
* #310: Add Extension:CollapsibleVector
* #300: Improve landing page
* #302: Email now functions
* #267: Address concerns about Digital Ocean installs failing (likely firewall issues)
* #278, #272, : Disable Extension:Flow
* #138: Cleanup apache.sh
* #175: Fix apache logrotate failures
* #342, #150, #151, #290: Handle custom extensions outside meza-core
* #369: Fixed admin panel on WikiBlender
* #367: Remove Extension:AccessDenied from saml.sh
* #279: SERVER_NAME not defined during maintenance scripts
* #309, #344: UploadWizard feedback page
* #329, #331: "Fix" image display in VE
* #306: Install Vim
* #353, #349: Fix httpd.conf broken symlink
* #373, #251: Improve UploadWizard license settings for enterprise environment
* #370, #378: Use specific version of SemanticForms (e.g. tags/3.5). This should be a pretty easy fix.
* #371: Move WikiBlender libraries into repo (don't use CDN). Also, don't show page views if they're not available. This isn't really a meza fix, and is just being tracked here.

The following issues are for information only:

* #368: Use this issue to track things that should be verified post-user-unify
* #149: Tracking progress in properly dealing with "image list" functionality. Parsoid is currently broken, and requires a bugfix to make external images function in VisualEditor. Unfortunately it is low-priority. This issue describes a workaround for the short-term.
* #297: A PR which was not merged (was closed), but was the original PR that was going to merge a lot of this content into master. It has much discussion about user-unify.

darenwelsh commented on .gitignore:
```diff
@@ -3,4 +3,13 @@ sources/*
 !sources/README.md
```
Is the purpose of this gitignore file to list what to ignore (and what to not ignore within some of those directories) for the entire meza installation? Does this enable someone to `git pull` and `git checkout` a new release as long as that release doesn't require a modified build from source or some other major change? Could you add some kind of introductory comment explaining this. I'm not really sure what to review in this file.

jamesmontalvo3 replied:
gitignore just specifies what files will or will not be added to your git repository when you do `git add -A`. For example, in MediaWiki if you do `git add -A` in the MediaWiki install directory, it won't add your `LocalSettings.php` file to the repository because the `.gitignore` file says to ignore that file. Now, you could force your repository to add that particular file by doing `git add LocalSettings.php`, but that's another story.

In meza there are some instances where all files in a directory are ignored (like `sources/*`), but a specific file is explicitly un-ignored (`!sources/README.md`). Basically the reason for this README is to be a placeholder for that directory so git tracks its existence even when nothing else is in it. When meza is first pulled, there is nothing in `sources` (other than the README), but the directory is a useful place to download stuff to during install.

darenwelsh commented on README.md:
```diff
@@ -1,4 +1,4 @@
-# meza v0.4
+# meza v0.9-dev
```
Why add the "-dev" to the release name?

jamesmontalvo3 replied:
Seemed like a good idea since this was such a long-running release. Didn't want the REL0_9 to seem like it was the stable released branch. It should be changed prior to merge.

darenwelsh commented on config/core/LocalSettings.php:
```diff
@@ -1076,71 +1173,71 @@
  **/
-if ( file_exists( "$m_htdocs/wikis/$wikiId/config/overrides.php" ) ) {
-	require_once "$m_htdocs/wikis/$wikiId/config/overrides.php";
+if ( file_exists( "$m_config/local/postLocalSettings_allWikis.php" ) ) {
```
Where/How does this file get generated? I see `preLocalSettings_allWikis.php` in meza/config/template, so should this directory also include a template for "post" settings modifications?

jamesmontalvo3 replied:
Off the top of my head I don't think this is generated anywhere, and it doesn't need to be. It's for customization, if desired (hence `if( file_exists( ... ) )`). 

Anything that is specified in `LocalSettings.php` would override anything specified in `preLocalSettings_allWikis.php`. Things not specified in `LocalSettings.php` that are specified in `preLocalSettings_allWikis.php` would not be overridden, however. So `preLocalSettings_allWikis.php` may be useful in some cases.

darenwelsh commented on config/core/LocalSettings.php:
```diff
@@ -263,18 +258,52 @@
+ *  In order for this to work properly the wikis need to have been created with
+ *  a single user table in mind. If you're starting a new wiki farm then you're
+ *  all set. If you're importing wikis which didn't previously have shared user
+ *  tables, then you'll need to use TBD user-merge script.
```
This TBD user-merge script works now, right?

jamesmontalvo3 replied:
yes this should call out the actual script name.

darenwelsh commented on config/core/LocalSettings.php:
```diff
@@ -452,14 +485,99 @@
+    $wgGroupPermissions['*']['read'] = true;
+    $wgGroupPermissions['user']['read'] = true;
+
+    // allow anonymous write
```
this should be "allow users write" or something meaning the same thing, right?

jamesmontalvo3 replied:
No, this is allowing anyone (anonymous or users) to edit.

darenwelsh replied:
I tagged my comment to the wrong line. Shouldn't this comment (line 515) be "allow only users write" or something meaning the same thing? Line 516 does not allow anyone to edit. Line 517 allows users to edit.

jamesmontalvo3 replied:
Good call. Fixing.

darenwelsh commented on manual/AddingWikis.md:
```diff
@@ -63,7 +63,7 @@ eng
 		favicon.ico
-		setup.php
+		preLocalSettings.php
 		CustomSettings.php
```
Does `CustomSettings.php` still exist? This PR includes a renamed file: `wiki-init/config/CustomSettings.php → ...te/wiki-init/config/postLocalSettings.php` which makes me think this should be updated.

jamesmontalvo3 replied:
Concur.

darenwelsh commented on manual/AddingWikis.md:
```diff
@@ -74,7 +74,7 @@ The contents of `config` are as follows:
 * `favicon.ico` is the wikis [favicon](https://en.wikipedia.org/wiki/Favicon).
-* `setup.php` has some meza-specific configuration variables in it. If you don't already have this file, let the install process generate one for you.
+* `preLocalSettings.php` has some meza-specific configuration variables in it. If you don't already have this file, let the install process generate one for you.
 * `CustomSettings.php` will also be auto-generated. It contains any settings specific to this particular wiki, different from your other wikis. Ideally it should be blank (keep all your wikis configured the same way).
```
Same comment as above -- does CustomSettings.php still exist?

darenwelsh commented on scripts/import-wikis.sh:
```diff
@@ -98,7 +107,7 @@ skipped_wikis=""
 #     images[/|.tar|.tar.gz]
 #     wiki.sql
-#     config/ (optional logo.png, favicon.ico, setup.php, CustomSettings.php)
+#     config/ (optional logo.png, favicon.ico, preLocalSettings.php, CustomSettings.php)
```
Same comment as in another file - does CustomSettings.php still exist?

jamesmontalvo3 replied:
No, fixing.

darenwelsh commented on scripts/extensions.sh:
```diff
@@ -57,7 +59,7 @@ echo -e "\n\n## meza: Install composer-supported extensions"
 cmd_profile "START extensions composer require"
 composer require \
-	mediawiki/semantic-media-wiki:~2.0 \
+	mediawiki/semantic-media-wiki:dev-master#787d70fb7f1c1db800e02e0dde0eee8ccc4b7e95 \
```
I think this is for #362 but I can't tell what the problem is that this fixes or why we can't check out 2.4 or some other actual release.

jamesmontalvo3 replied:
Can't checkout 2.4 because it's not released yet. We'd have to be running on SMW's master branch, which is a moving target. Regarding why we need this commit: I did a pretty crappy job documenting, didn't I? I've updated #362 with a little more info. Basically: `rebuildData.php` fails on some pages without this.

darenwelsh commented on scripts/install.sh:
```diff
@@ -13,6 +13,8 @@ if [[ $PATH != *"/usr/local/bin"* ]]; then
 
+working_dir=`pwd`
+
 echo -e "\nWelcome to meza v0.4\n"
```
should this be v0.9?

jamesmontalvo3 replied:
Yes

jamesmontalvo3 commented:
I just updated this PR's summary with references to relevant issues and PRs. I'll add to the "changes" and "testing" sections later.

jamesmontalvo3 commented:
Added list of changes to PR's summary. There is probably some overlap with the list of addressed issues and PRs, but I figured hitting all the points was important.

jamesmontalvo3 commented:
Updated "testing" section of PR summary

jamesmontalvo3 commented:
Changes made today: bump SemanticForms to version 3.5 rather than just tracking the REL1_25 branch. I don't recall exactly why this was required, but it was on the 0.9 list. Also did a hotfix for the EPEL repository URL being broken.

jamesmontalvo3 commented:
After much thought, we've decided to merge this into master. It's a massive commit, and our testing  indicates that everything is working. The true test with such a complex system will be getting lots of users to use it.
